### PR TITLE
[quantization] FP16 dynamic quantized Linear

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32331 [quantization] FP16 dynamic quantized Linear**

Differential Revision: [D19441158](https://our.internmc.facebook.com/intern/diff/D19441158)

### CODE REVIEW

File: test/test_quantization.py [hunk 1]
```diff
@@ -319,187 +320,195 @@ def test_single_layer(self):
...
-        model = SingleLayerLinearDynamicModel()
-        quantize_dynamic(model, qconfig_dict, inplace=True)
-        checkQuantized(model)
-
-        # Test set qconfig
-        model = SingleLayerLinearDynamicModel()
-        quantize_dynamic(model, set([nn.Linear]), inplace=True)
-        checkQuantized(model)
+        # for dtype in [torch.qint8, torch.float16]:
+        for dtype in [torch.float16]:
```
Comment (raghuramank100) [inline test/test_quantization.py]: Why is torch.qint8 excluded in the test?

File: torch/nn/quantized/modules/linear.py [hunk 2]
```diff
@@ -30,13 +47,21 @@ def _save_to_state_dict(self, destination, prefix, keep_vars):
...
         destination[prefix + 'bias'] = b
+        destination[prefix + 'dtype'] = self.dtype
 
     def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
                               missing_keys, unexpected_keys, error_msgs):
         self.set_weight_bias(state_dict[prefix + 'weight'], state_dict[prefix + 'bias'])
         state_dict.pop(prefix + 'weight')
         state_dict.pop(prefix + 'bias')
 
+        version = local_metadata.get('version', None)
```
Comment (raghuramank100) [inline torch/nn/quantized/modules/linear.py]: Do we need to serialize dtype? This is an attribute that is defined in the code at init.

File: torch/nn/quantized/modules/linear.py [hunk 3]
```diff
@@ -47,11 +72,11 @@ def __getstate__(self):
                                ' See https://github.com/pytorch/pytorch/issues/24045.'
                                ' Please use state_dict or torch.jit serialization.')
         qweight, bias = self._weight_bias()
-        return qweight, bias, self.training
+        return qweight, bias, self.training, self.dtype
```
Comment (raghuramank100) [inline torch/nn/quantized/modules/linear.py]: Do we still need this warning? The github issue is closed now.

File: torch/nn/quantized/modules/linear.py [hunk 4]
```diff
@@ -8,19 +8,36 @@
 from torch.nn.quantized.modules.utils import _quantize_weight
 
 class LinearPackedParams(torch.nn.Module):
-    def __init__(self):
+    _version = 2
```
Comment (raghuramank100) [inline torch/nn/quantized/modules/linear.py]: Not sure if _version is needed as we do not need to have dtype be part of the state dict.

File: torch/nn/quantized/modules/linear.py [hunk 5]
```diff
@@ -81,10 +106,10 @@ class Linear(torch.nn.Module):
         >>> print(output.size())
         torch.Size([128, 30])
     """
-    _version = 2
+    _version = 3
```
Comment (raghuramank100) [inline torch/nn/quantized/modules/linear.py]: Can you add a comment explaining why the version has changed? I see no change in the class that uses the new version number.

**Review by raghuramank100 — DISMISSED** [review]
Looks great, a few questions/comments.

File: torch/nn/quantized/modules/linear.py (hunk 5 above)
Comment (jamesr66a) [reply torch/nn/quantized/modules/linear.py]: In general you're never going to see the current version number: the straightline control flow of the program is the "current" version. You'll only see branches on previous version numbers to adapt some changed functionality

File: torch/nn/quantized/modules/linear.py (hunk 2 above)
Comment (jamesr66a) [reply torch/nn/quantized/modules/linear.py]: Yes. `dtype` is branched on in `set_weight_bias` which is called from the deserialization methods

File: torch/nn/quantized/modules/linear.py (hunk 3 above)
Comment (jamesr66a) [reply torch/nn/quantized/modules/linear.py]: Not sure what you're referring to, https://github.com/pytorch/pytorch/issues/24045 is not closed

File: torch/nn/quantized/modules/linear.py (hunk 4 above)
Comment (jamesr66a) [reply torch/nn/quantized/modules/linear.py]: See below, `dtype` is definitely needed

File: test/test_quantization.py (hunk 1 above)
Comment (jamesr66a) [reply test/test_quantization.py]: Leftover from debugging, good catch!

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 31094700:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=02aa3ba3312c273472084f782f06f79109185f33&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 02aa3ba3 since Jan 23
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=3109470053bbc704d35db080137d6eff183ff163).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=89874920) from Jan 23



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/nn/quantized/modules/linear.py (hunk 3 above)
Comment (raghuramank100) [reply torch/nn/quantized/modules/linear.py]: Sorry, was referring to this one https://github.com/pytorch/pytorch/pull/29232

**Review by raghuramank100 — APPROVED** [review]

**Pull request closed without merging**